### PR TITLE
Make packer pack smaller granularity and round-robin

### PR DESCRIPTION
**Pull request by Jackmin801**
The current code has an issue where very big batches being completed by an orchestrator will suddenly flood the trainer with work. This change should help mitigate that by making the steps smaller and thus, the loop checks for if it should step optimizer and broadcast more often, allowing smaller runs to be interleaved with the the processing of a big run.

Also added tests as Jannik seems to be reporting issues with the packer giving a rank empty microbatch. These tests should be able to catch what's wrong

---

> [!NOTE]
> Implements fair, small‑granularity packing and robust run lifecycle handling for multi‑run RL training.
> 
> - Introduces `BasePacker` with `SinglePacker` and `MultiPacker`; `MultiPacker` buffers `TrainingSample`s per run, selects samples via round‑robin up to `seq_len * dp_world_size`, and updates per‑run progress only when `batch_size` is met
> - Adds `setup_packer` and integrates it in `DataLoader` (master only), replacing direct `Packer` usage and renaming `config`→`transport_config`
> - Enhances run management with `register_create_run_data_hook`/`register_delete_run_data_hook` in `Runs`, called on run add/remove; `MultiPacker` clears buffers and resets receiver on delete
> - Transport: add `reset_run()` to `TrainingBatchReceiver`; filesystem receiver now tracks per‑run `_received_steps` to prevent duplicate reads and re‑initializes on reset
> - Minor cleanup/formatting in training loop; CI trainer config increases `max_steps` for integration test
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 003ffb769d28709da377376ae4607fe81f197dfc. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/rl/packer.py [hunk 1]
```diff
@@ -27,70 +28,174 @@ def __init__(
...
-            self.runs.progress[idx].step += 1
-            self.runs.progress[idx].total_tokens += sum(
-                len(rollout.prompt_ids) + len(rollout.completion_ids) for rollout in training_batch.examples
-            )
-            self.runs.progress[idx].total_samples += len(training_batch.examples)
-            self.runs.ready_to_update[idx] = True
+
+        for run_idx, sample_temp_pairs in samples_by_run.items():
+            samples = [s for s, _ in sample_temp_pairs]
+            temperature = sample_temp_pairs[0][1]
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/packer.py:224]: this doesnt matter because we dont ever change the temperature in the orchestrator atm

File: src/prime_rl/trainer/rl/config.py [hunk 2]
```diff
@@ -84,6 +84,13 @@ class DataLoaderConfig(BaseConfig):
     """Configures the data loader used for training."""
 
     fake: Annotated[FakeDataLoaderConfig | None, Field(description="Whether to use a fake data loader.")] = None
+    small_batch_granularity: Annotated[
```
Comment (samsja) [inline src/prime_rl/trainer/rl/config.py]: not sure about the naming here, not super clear what it does from the name

**Review by samsja — COMMENTED** [review]
I struggle to understand if the logic is only for multi run or for any rl run, in the case of any rl run I don't understand what it does. Maybe we should seperate the packer for single and multi lora run ?

File: src/prime_rl/trainer/rl/config.py (hunk 2 above)
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/config.py]: renamed to `pack_only_one_microbatch_in_each_step`. should be descriptive enough

File: src/prime_rl/trainer/rl/packer.py [hunk 3]
```diff
@@ -40,58 +46,218 @@ def __init__(
...
+                    return selected
+                selected.append((run_idx, sample, temperature))
+                self.buffers[run_idx].popleft()
+
+        return selected
+
+    def _update_run_progress(self, run_idx: int, num_samples: int, num_tokens: int) -> None:
+        """Update progress, increment step only when batch_size reached."""
+        self.samples_consumed_this_step[run_idx] += num_samples
+        batch_size = self.runs.config[run_idx].batch_size
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/packer.py:184]: yea good catch, should have a delete hook instead

File: src/prime_rl/trainer/rl/packer.py [hunk 4]
```diff
@@ -40,58 +46,218 @@ def __init__(
...
+        while len(batches) == 0:
+            self.runs.check_for_changes()
+            batches = self.receiver.receive()
+            time.sleep(0.2)
+
+        assert len(batches) == 1, "SinglePacker only supports one batch per step"
+        batch = batches[0]
+
+        self.runs.ready_to_update[0] = True
+        self.runs.progress[0].step += 1
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/packer.py:76]: I dont think they get used anywhere so this is probably fine

File: src/prime_rl/trainer/rl/packer.py [hunk 5]
```diff
@@ -40,58 +42,218 @@ def __init__(
...
+
+            while len(self.buffers[run_idx]) > 0:
+                sample, temperature = self.buffers[run_idx][0]
+                tokens_collected += len(sample.prompt_ids) + len(sample.completion_ids)
+                if tokens_collected > token_budget:
+                    if tokens_collected == (len(sample.prompt_ids) + len(sample.completion_ids)):
+                        tokens_collected -= len(sample.prompt_ids) + len(sample.completion_ids)
+                        # This means we have a sample that has more tokens than max seqlen
+                        self.buffers[run_idx].popleft()
+                        continue
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/packer.py:174]: aaaaaaah whatever lets assume it doesnt happen for now

**Review by samsja — APPROVED** [review]
